internal/cmd/login.go

```diff
@@ -1,83 +1,149 @@
-		Scheme:   u.Scheme,
-		Host:     u.Host,
-		Path:     "/internal-auth/",
-		RawQuery: "local_service=http://" + listener.Addr().String(),
-	}
+func login(cmd *cobra.Command, envURL *url.URL, urlCfg, sessionCfg config.File) error {
+	ctx := cmd.Context()
 
-	err = browser.OpenURL(authURL.String())
+	// Start by creating the listener so we can prompt the user with the URL.
+	listener, err := newLocalListener()
 	if err != nil {
+		return xerrors.Errorf("create local listener: %w", err)
+	}
+	defer func() { _ = listener.Close() }() // Best effort.
+
+	// Forge the auth URL with the callback set to the local server.
+	authURL := *envURL
+	authURL.Path = envURL.Path + "/internal-auth"
+	authURL.RawQuery = "local_service=http://" + listener.Addr().String()
+
+	// Try to open the browser on the local computer.
+	if err := browser.OpenURL(authURL.String()); err != nil {
+		// Discard the error as it is an expected one in non-X environments like over ssh.
 		// Tell the user to visit the URL instead.
-		flog.Info("visit %s to login", authURL.String())
+		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Visit the following URL in your browser:\n\n\t%s\n\n", &authURL) // Can't fail.
 	}
-	srv.TokenCond.L.Lock()
-	srv.TokenCond.Wait()
-	err = config.Session.Write(srv.Token)
-	srv.TokenCond.L.Unlock()
-	if err != nil {
-		return xerrors.Errorf("set session: %v", err)
+
+	// Create our channel, it is going to be the central synchronization of the command.
+	tokenChan := make(chan string)
+
+	// Create the http server outside the errgroup goroutine scope so we can stop it later.
+	srv := &http.Server{Handler: &loginsrv.Server{TokenChan: tokenChan}}
+	defer func() { _ = srv.Close() }() // Best effort. Direct close as we are dealing with a one-off request.
+
+	// Start both the readline and http server in parallel. As they are both long-running routines,
+	// to know when to continue, we don't wait on the errgroup, but on the tokenChan.
+	group, ctx := errgroup.WithContext(ctx)
+	group.Go(func() error { return srv.Serve(listener) })
+	group.Go(func() error { return loginsrv.ReadLine(ctx, cmd.InOrStdin(), cmd.ErrOrStderr(), tokenChan) })
+
+	// Only close then tokenChan when the errgroup is done. Best effort basis.
+	// Will not return the http route is used with a regular terminal.
+	// Useful for non interactive session, manual input, tests or custom stdin.
+	go func() { defer close(tokenChan); _ = group.Wait() }()
+
+	var token string
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case token = <-tokenChan:
+	}
+
+	// Perform an API call to verify that the token is valid.
+	if err := pingAPI(ctx, envURL, token); err != nil {
+		return xerrors.Errorf("ping API: %w", err)
 	}
-	flog.Success("logged in")
+
+	// Success. Store the config only at this point so we don't override the local one in case of failure.
+	if err := storeConfig(envURL, token, urlCfg, sessionCfg); err != nil {
+		return xerrors.Errorf("store config: %w", err)
+	}
+
+	flog.Success("Logged in.")
```

Comment:
Should be precede this with a newline? Right now, the output is:
```
$ coder login https://master.cdr.dev
or enter token manually: 2020-09-08 16:04:30 SUCCESS    Logged in.
```

Reply:
Added a newline in the prompt before this line. Otherwise, flog's prefix still is on he previous line.